### PR TITLE
Update closure capture error logging for disjoint captures for disjoint captures

### LINKED ISSUE #8: Update error logging in borrow code
Depends on #22 and #10

### LINKED ISSUE #36: Print captures precisely in borrow checker
The borrow checker just prints the root variable when printing captures from within the context of the closure, we should update this to be print the captured path precisely.

Ideally, we do this once rust-lang/rust#80635 lands because it introduces a function to prints paths precisely.

### LINKED ISSUE #39: Incorrect diagnostic when the feature isn't enabled
Compiling tte following code using the old capture logic
```rust
#![feature(rustc_attrs)]

struct S(i32);

fn foo() {
    let b = Box::new(S(10));
    
    let c = #[rustc_capture_analysis] || {
        let _b = b.0;
    };
    
    c();
}
```

prints the following diagnostic when compiled([view on playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1f4748becb6031c1ffaf47c2175893a3)):
```
note: Min Capture b[] -> ImmBorrow
  --> src/lib.rs:9:39
   |
9  |       let c = #[rustc_capture_analysis] || {
   |  _______________________________________^
10 | |         let _b = b.0;
   | |                  ^^^ b[] captured as ImmBorrow here
11 | |     };
   | |_____^ b[] used here
```
It should instead look something like:
```
note: Min Capture b[] -> ImmBorrow
  --> /users/cpardy/example2.rs:10:18
   |
10 |         let _b = b.0;
   |                  ^^^
```

### LINKED ISSUE #76005: Compute `Upvar.span` after typecheck
In https://github.com/rust-lang/rust/pull/75933, I added code to determine a better upvar span when emitting borrow check errors. This span is stored in `ty::UpvarCapture`.

Ideally, we would return this span from the `upvars_mentioned` query - however, the better span depends on typeck (so we can tell if the upvar is inferred to be by-value), and typeck uses `upvars_mentioned`.

### PR DESCRIPTION
**Pull request by ChrisPardy**
Improved error logging when `#![feature(capture_disjoint_fields)]` is used.

Closes https://github.com/rust-lang/project-rfc-2229/issues/8
Closes https://github.com/rust-lang/project-rfc-2229/issues/36
Closes https://github.com/rust-lang/project-rfc-2229/issues/39
Closes #76005

### CODE REVIEW

File: compiler/rustc_middle/src/mir/mod.rs [hunk 1]
```diff
@@ -2346,6 +2355,7 @@ impl<'tcx> Debug for Rvalue<'tcx> {
                             };
                             let mut struct_fmt = fmt.debug_struct(&name);
 
+                            // FIXME: This should be a list of capture names/places
```
Comment (nikomatsakis) [inline compiler/rustc_middle/src/mir/mod.rs:2358]: Is there an issue number for this FIXME?

File: compiler/rustc_mir/src/borrow_check/diagnostics/explain_borrow.rs [hunk 2]
```diff
@@ -67,22 +67,39 @@ impl BorrowExplanation {
...
                     LaterUseKind::Other => "used here",
                 };
-                if !borrow_span.map_or(false, |sp| sp.overlaps(var_or_use_span)) {
-                    err.span_label(
-                        var_or_use_span,
-                        format!("{}borrow later {}", borrow_desc, message),
-                    );
+                // We can use `var_or_use_span` if either `path_span` is not present, or both spans are the same
+                if path_span.map_or(true, |path_span| path_span == var_or_use_span) {
+                    if !borrow_span.map_or(false, |sp| sp.overlaps(var_or_use_span)) {
```
Comment (nikomatsakis) [inline compiler/rustc_mir/src/borrow_check/diagnostics/explain_borrow.rs:80]:
```suggestion
                if path_span.map(|path_span| path_span == var_or_use_span).unwrap_or(true) {
                    if borrow_span.map(|sp| !sp.overlaps(var_or_use_span)).unwrap_or(true) {
```

I have a deep, abiding, and personal hatred for `map_or` and I am going to impose it on you.

File: compiler/rustc_mir/src/borrow_check/diagnostics/explain_borrow.rs [hunk 3]
```diff
@@ -94,7 +111,24 @@ impl BorrowExplanation {
                     LaterUseKind::FakeLetRead => "borrow later stored here",
                     LaterUseKind::Other => "borrow used here, in later iteration of loop",
                 };
-                err.span_label(var_or_use_span, format!("{}{}", borrow_desc, message));
+                // We can use `var_or_use_span` if either `path_span` is not present, or both spans are the same
+                if path_span.map_or(true, |path_span| path_span == var_or_use_span) {
```
Comment (nikomatsakis) [inline compiler/rustc_mir/src/borrow_check/diagnostics/explain_borrow.rs:115]:
```suggestion
                if path_span.map(|path_span| path_span == var_or_use_span).unwrap_or(true) {
```
As above =)

File: compiler/rustc_mir/src/borrow_check/diagnostics/explain_borrow.rs [hunk 4]
```diff
@@ -94,7 +111,24 @@ impl BorrowExplanation {
...
+                    err.span_label(var_or_use_span, format!("{}{}", borrow_desc, message));
+                } else {
+                    // path_span must be `Some` as otherwise the if condition is true
+                    let path_span = path_span.unwrap();
+                    // path_span is only present in the case of closure capture
+                    assert!(matches!(later_use_kind, LaterUseKind::ClosureCapture));
+                    if !borrow_span.map_or(false, |sp| sp.overlaps(var_or_use_span)) {
```
Comment (nikomatsakis) [inline compiler/rustc_mir/src/borrow_check/diagnostics/explain_borrow.rs:122]:
```suggestion
                    if borrow_span.map(|sp| !sp.overlaps(var_or_use_span)).unwrap_or(true) {
```

File: compiler/rustc_mir/src/borrow_check/diagnostics/mod.rs [hunk 5]
```diff
@@ -600,9 +602,21 @@ impl UseSpans<'_> {
         }
     }
 
+    pub(super) fn var_or_use_path_span(self) -> Span {
```
Comment (nikomatsakis) [inline compiler/rustc_mir/src/borrow_check/diagnostics/mod.rs:605]: A doc-comment would I think be nice here

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks pretty good! Let a few nits, but r=me once fixed.

**Comment by arora-aman** [comment]
Closes #76005 aswell

File: compiler/rustc_mir/src/borrow_check/diagnostics/region_errors.rs [hunk 6]
```diff
@@ -385,6 +385,7 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
 
         diag.span_label(*span, message);
 
+        // FIXME: This should store a captured_place not a hir id
```
[the remark this answers is not in the document]
Comment (ChrisPardy) [reply compiler/rustc_mir/src/borrow_check/diagnostics/region_errors.rs]: I've created https://github.com/rust-lang/project-rfc-2229/issues/48 for the FIXMEs created in this PR

**Pull request merged**